lib/fauna/page.rb

```diff
@@ -0,0 +1,381 @@
+    attr_reader :before
+
+    ##
+    # After cursor for the current page.
+    #
+    # Always +nil+ for the initial page. Call one of the pagination methods to begin paging.
+    attr_reader :after
+
+    # :section: Builders
+
+    ##
+    # Returns a copy of the page with the given +ts+ set.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_ts(ts)
+      with_dup do |page|
+        page.instance_variable_get(:@page_params)[:ts] = ts
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given cursor set.
+    #
+    # Cursors either contain a +before+ or +after+ parameter.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_cursor(cursor = {})
+      with_dup do |page|
+        page.send(:load_cursor, cursor)
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given +size+ set.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_size(size = nil)
+      with_dup do |page|
+        page.instance_variable_get(:@page_params)[:size] = size
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given +events+ set.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_events(events)
+      with_dup do |page|
+        page.instance_variable_get(:@page_params)[:events] = events
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given +sources+ set.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_sources(sources)
+      with_dup do |page|
+        page.instance_variable_get(:@page_params)[:sources] = sources
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given fauna block set.
+    #
+    # The block, when provided, will be used to wrap the generated paginate query with a fauna query.
+    # The block will be run in a Query.expr context, and passed the generated paginate query as a parameter.
+    #
+    # Example of mapping a set of refs to their instances:
+    #
+    #   page.with_fauna_map { |page_q| map(page_q) { |ref| get ref } }
+    def with_fauna_map(&block)
+      with_dup do |page|
+        page.instance_variable_set(:@fauna_map, block)
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given ruby block set.
+    #
+    # The block, when provided, will be used to map the returned data elements from the executed query.
+    # The block will be passed the each element as a parameter from the data of the page currently being loaded.
+    #
+    # Example of loading instances into your own model:
+    #
+    #   page.with_ruby_map { |instance| YourModel.load(instance) }
+    def with_ruby_map(&block)
```

Comment:
...The more I think about it, this seems a bit confusing. would it be possible to just map the returned data or enumerator?